packages/peering/lpp/lppPeer.go

```diff
@@ -0,0 +1,219 @@
+// Copyright 2020 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+package lpp
+
+import (
+	"sync"
+	"time"
+
+	"github.com/iotaledger/hive.go/crypto/ed25519"
+	"github.com/iotaledger/hive.go/logger"
+	"github.com/iotaledger/wasp/packages/peering"
+	libp2ppeer "github.com/libp2p/go-libp2p-core/peer"
+	"go.uber.org/atomic"
+	"golang.org/x/xerrors"
+)
+
+const (
+	inactiveDeadline = 1 * time.Minute
+	inactivePingTime = 30 * time.Second
+)
+
+type peer struct {
+	remoteNetID    string
+	remotePubKey   *ed25519.PublicKey
+	remoteLppID    libp2ppeer.ID
+	accessLock     *sync.RWMutex
+	sendCh         chan *peering.PeerMessage
+	sendChOverflow atomic.Uint32
+	lastMsgSent    time.Time
+	lastMsgRecv    time.Time
+	numUsers       int
+	trusted        bool
+	net            *netImpl
+	log            *logger.Logger
+}
+
+func newPeer(remoteNetID string, remotePubKey *ed25519.PublicKey, remoteLppID libp2ppeer.ID, n *netImpl) *peer {
+	log := n.log.Named("peer:" + remoteNetID)
+	p := &peer{
+		remoteNetID:  remoteNetID,
+		remotePubKey: remotePubKey,
+		remoteLppID:  remoteLppID,
+		accessLock:   &sync.RWMutex{},
+		sendCh:       make(chan *peering.PeerMessage, 100),
+		lastMsgSent:  time.Time{},
+		lastMsgRecv:  time.Time{},
+		numUsers:     0,
+		trusted:      true,
+		net:          n,
+		log:          log,
+	}
+	go p.sendLoop()
+	return p
+}
+
+func (p *peer) usePeer() {
+	p.accessLock.Lock()
+	defer p.accessLock.Unlock()
+	p.numUsers++
+}
+
+func (p *peer) noteReceived() {
+	p.accessLock.Lock()
+	defer p.accessLock.Unlock()
+	p.lastMsgRecv = time.Now()
+}
+
+// Send pings, if needed. Other periodic actions can be added here.
+func (p *peer) maintenanceCheck() {
+	now := time.Now()
+	old := now.Add(-inactivePingTime)
+	p.accessLock.RLock()
+	if p.numUsers > 0 && p.lastMsgRecv.Before(old) {
+		p.accessLock.RUnlock()
+		p.net.lppHeartbeatSend(p, true)
+	} else {
+		p.accessLock.RUnlock()
+	}
+	if p.numUsers == 0 && !p.trusted && p.lastMsgRecv.Before(old) {
+		p.net.delPeer(p)
+		close(p.sendCh)
+	}
+}
```

Comment:
this conditional unlock logic is weird.
```
	p.accessLock.RLock()
	if p.numUsers > 0 && p.lastMsgRecv.Before(old) {
		p.accessLock.RUnlock()
		p.net.lppHeartbeatSend(p, true)
	} else {
		p.accessLock.RUnlock()
	}
```

shouldn't we also lock/unlock `p.accessLock` to read on line #80 ?
```
	if p.numUsers == 0 && !p.trusted && p.lastMsgRecv.Before(old) {
```

what do you think about the following?
```suggestion
func (p *peer) maintenanceCheck() {
	now := time.Now()
	old := now.Add(-inactivePingTime)
	
	p.accessLock.RLock()
	numUsers := p.numUsers
	isLastMsgOld := p.lastMsgRecv.Before(old)
	trusted := p.trusted
	p.accessLock.RUnlock()
	
	if numUsers > 0 &&  isLastMsgOld {
		p.net.lppHeartbeatSend(p, true)
	}
	if numUsers == 0 && !trusted && isLastMsgOld {
		p.net.delPeer(p)
		close(p.sendCh)
	}
}
```

Reply:
Nice pattern. Will use it.